### PR TITLE
MULE-18926: Decouple ComponentBuildingDefinitionRegistry creation from MuleArtifactContext

### PR DESCRIPTION
**Pull request by mlischetti**

### CODE REVIEW

**Comment by mlischetti** [comment]
--validate

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DefaultApplicationFactory.java [hunk 1]
```diff
@@ -125,6 +126,12 @@ public ApplicationDescriptor createArtifactDescriptor(File artifactLocation, Opt
   }
 
   public Application createArtifact(ApplicationDescriptor descriptor) throws IOException {
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DefaultApplicationFactory.java:128]: should this be deprecated? are there any remaining usages of this method?

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/ArtifactContextBuilder.java [hunk 2]
```diff
@@ -374,6 +376,18 @@ public ArtifactContextBuilder setRuntimeLockFactory(LockFactory runtimeLockFacto
     return this;
   }
 
+  /**
+   * Allows the usage of {@link ComponentBuildingDefinitionRegistryFactory} to create a {@link org.mule.runtime.dsl.api.component.ComponentBuildingDefinitionRegistry}.
+   * 
+   * @param componentBuildingDefinitionRegistryFactory the {@link ComponentBuildingDefinitionRegistryFactory} factory
+   *                                                   used to create a {@link org.mule.runtime.dsl.api.component.ComponentBuildingDefinitionRegistry}
+   * @return
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/ArtifactContextBuilder.java:384]: missing description

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/domain/DefaultDomainFactory.java [hunk 3]
```diff
@@ -134,6 +134,12 @@ public void start(DomainWrapper domainWrapper) {
 
   @Override
   protected Domain doCreateArtifact(File domainLocation, Optional<Properties> deploymentProperties) throws IOException {
+    return createArtifact(domainLocation, deploymentProperties, null);
+  }
+
+  public Domain createArtifact(File domainLocation, Optional<Properties> deploymentProperties,
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/domain/DefaultDomainFactory.java:140]: what's the purpose of this split? where is this new method called?

File: modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleCallingOperationsWithinModuleTestCase.java [hunk 4]
```diff
@@ -195,4 +195,9 @@ private CoreEvent runFlowWithContentAndDefaultVariables(String flowName) throws
         .withVariable("secondPart", SECOND_PART)
         .run();
   }
+
+  @Override
+  public boolean mustRegenerateExtensionModels() {
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleCallingOperationsWithinModuleTestCase.java:200]: add a comment explaining why this is needed

File: modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleGlobalElementTestCase.java [hunk 5]
```diff
@@ -87,5 +87,9 @@ private void assertGetClient(String flow) throws Exception {
     assertThat(client.getPassword(), is("notDoe"));
   }
 
+  @Override
+  public boolean mustRegenerateExtensionModels() {
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleGlobalElementTestCase.java:91]: add a comment explaining why this is needed

File: modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleSimpleTestCase.java [hunk 6]
```diff
@@ -202,4 +202,9 @@ public void testSetPayloadHardcodedFlowThruSubflowWithNestedElements() throws Ex
     CoreEvent event = flowRunner("testSetPayloadHardcodedFlowThruSubflowWithNestedElements").run();
     assertThat(event.getMessage().getPayload().getValue(), is("hardcoded value"));
   }
+
+  @Override
+  public boolean mustRegenerateExtensionModels() {
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleSimpleTestCase.java:207]: add a comment explaining why this is needed

File: modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleTestConnectionTestCase.java [hunk 7]
```diff
@@ -71,4 +71,9 @@ private void assertConnectionOn(String beanName) throws MuleException {
     assertThat(connectionValidationResult.isValid(), is(true));
     connectionProvider.disconnect(connect);
   }
+
+  @Override
+  public boolean mustRegenerateExtensionModels() {
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/test/java/org/mule/test/functional/ModuleTestConnectionTestCase.java:76]: add a comment explaining why this is needed

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/MuleArtifactContext.java [hunk 8]
```diff
@@ -227,21 +207,12 @@ public MuleArtifactContext(MuleContext muleContext, ConfigResource[] artifactCon
...
-
-    for (ClassLoader pluginArtifactClassLoader : pluginsClassLoaders) {
-      getArtifactComponentBuildingDefinitions(pluginArtifactClassLoader)
-          .forEach(componentBuildingDefinitionRegistry::register);
-    }
 
     this.beanDefinitionFactory =
-        new BeanDefinitionFactory(muleContext.getConfiguration().getId(), componentBuildingDefinitionRegistry);
+        new BeanDefinitionFactory(muleContext.getConfiguration().getId(),
+                                  componentBuildingDefinitionRegistryFactory.create(extensions));
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/MuleArtifactContext.java:215]: inline `extensions`

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/domain/DefaultDomainFactory.java (hunk 3 above)
Comment (mlischetti) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/domain/DefaultDomainFactory.java:140]: this should be used by tooling services

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DefaultApplicationFactory.java (hunk 1 above)
Comment (mlischetti) [reply modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DefaultApplicationFactory.java:128]: the unique consumer is tooling services and should use the new one. So this method can be deprecated

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**